src/mempoollimit.cpp

```diff
@@ -0,0 +1,159 @@
+    int rightWeight = getWeightAt(fromIndex).lowFeePenaltyWeight - getWeightAt(fromIndex * 2 + 2).lowFeePenaltyWeight;
+    // On Left
+    if (weightToFind < leftWeight) {
+        return findByWeight(fromIndex * 2 + 1, weightToFind);
+    }
+    // Found
+    if (weightToFind < rightWeight) {
+        return fromIndex;
+    }
+    // On Right
+    return findByWeight(fromIndex * 2 + 2, weightToFind - rightWeight);
+}
+
+TxWeight WeightedTxTree::getTotalWeight() const
+{
+    return getWeightAt(0);
+}
+
+
+void WeightedTxTree::add(const WeightedTxInfo& weightedTxInfo)
+{
+    txIdAndWeights.push_back(weightedTxInfo);
+    childWeights.push_back(ZERO_WEIGHT);
+    txIdToIndexMap[weightedTxInfo.txId] = size;
+    backPropagate(size++, weightedTxInfo.txWeight);
+}
+
+void WeightedTxTree::remove(const uint256& txId)
+{
+    if (txIdToIndexMap.find(txId) == txIdToIndexMap.end()) {
+        return;
+    }
+
+    size_t removeIndex = txIdToIndexMap[txId];
+
+    TxWeight lastChildWeight = txIdAndWeights[--size].txWeight;
+    backPropagate(size, lastChildWeight.negate());
+
+    if (removeIndex < size) {
+        TxWeight weightDelta = lastChildWeight.add(txIdAndWeights[removeIndex].txWeight.negate());
+        txIdAndWeights[removeIndex] = txIdAndWeights[size];
+        txIdToIndexMap[txIdAndWeights[removeIndex].txId] = removeIndex;
+        backPropagate(removeIndex, weightDelta);
+    }
+
+    txIdToIndexMap.erase(txId);
+    txIdAndWeights.pop_back();
+    childWeights.pop_back();
+}
+
+boost::optional<uint256> WeightedTxTree::maybeDropRandom()
+{
+    uint64_t totalPenaltyWeight = getTotalWeight().lowFeePenaltyWeight;
+    if (totalPenaltyWeight <= capacity) {
+        return boost::none;
+    }
+    LogPrint("mempool", "Mempool cost limit exceeded (cost=%d, limit=%d)\n", totalPenaltyWeight, capacity);
+    int randomWeight = GetRand(totalPenaltyWeight);
+    WeightedTxInfo drop = txIdAndWeights[findByWeight(0, randomWeight)];
+    LogPrint("mempool", "Evicting transaction (txid=%s, cost=%d, penaltyCost=%d)\n",
+        drop.txId.ToString(), drop.txWeight.weight, drop.txWeight.lowFeePenaltyWeight);
+    remove(drop.txId);
+    return drop.txId;
+}
+
+
+TxWeight TxWeight::add(const TxWeight& other) const
+{
+    return TxWeight(weight + other.weight, lowFeePenaltyWeight + other.lowFeePenaltyWeight);
+}
+
+TxWeight TxWeight::negate() const
+{
+    return TxWeight(-weight, -lowFeePenaltyWeight);
+}
+
+
+// These are also defined in rpcwallet.cpp
+#define JOINSPLIT_SIZE GetSerializeSize(JSDescription(), SER_NETWORK, PROTOCOL_VERSION)
+#define OUTPUTDESCRIPTION_SIZE GetSerializeSize(OutputDescription(), SER_NETWORK, PROTOCOL_VERSION)
+#define SPENDDESCRIPTION_SIZE GetSerializeSize(SpendDescription(), SER_NETWORK, PROTOCOL_VERSION)
+
+WeightedTxInfo WeightedTxInfo::from(const CTransaction& tx, const CAmount& fee)
+{
+    size_t memUsage = RecursiveDynamicUsage(tx);
+    memUsage += tx.vJoinSplit.size() * JOINSPLIT_SIZE;
```

Comment:
Should these instead be folded into `RecursiveDynamicUsage`?

Reply:
Should this be fixed as part of this issue, or can it be done as follow up?

Reply:
If we did this as follow up, we could also incorporate this in to `z_mergetoaddress` which does its own tx size estimation, but in a different way.